SU2_CFD/include/variables/CVariable.hpp

```diff
@@ -94,6 +94,101 @@ class CVariable {
   su2matrix<int> AD_InputIndex;    /*!< \brief Indices of Solution variables in the adjoint vector. */
   su2matrix<int> AD_OutputIndex;   /*!< \brief Indices of Solution variables in the adjoint vector after having been updated. */
 
+  VectorType SolutionExtra; // Stores the full adjoint solution
+  VectorType ExternalExtra; // External storage for the adjoint value (i.e. for the OF mainly)
+
+  VectorType SolutionExtra_BGS_k; // Intermediate storage, enables cross term extraction as that is also pushed to Solution
+
+  su2vector<int> AD_InputIndex_Extra; // Stores input index of the Extra vars
+  su2vector<int> AD_OutputIndex_Extra; // Stores output index of the Extra vars
+
+  VectorType SolutionExtra_Direct; // Stores values of the original direct solution, for later resetting
+  su2vector<su2double*> SolutionExtra_OriginAdresses; // Stores pointers to the original vars, these vars are resetted via this pointer vector
+
+ public: // Make public for easy debug access
+
+  // Add an su2double to the origin addresses for the extra solution variables
+  // only called by direct solver
+  inline void AddSolutionExtra_OriginAdress(su2double* val) {
+    SolutionExtra_OriginAdresses.resize(1); // here a push_back would be premium
+    SolutionExtra_OriginAdresses[0] = val;
+
+    AD_InputIndex_Extra.resize(SolutionExtra_OriginAdresses.size()) = -1;
+    AD_OutputIndex_Extra.resize(SolutionExtra_OriginAdresses.size()) = -1;
```

Comment:
This is still a hack for just 1 additional variable. With a push_back() this would be really nice -> The `In/Output_Index` allocation of course could move somewhere or I always resize to `size()+1` .... wait I could do that to `SolutionExtra_OriginAdresses` (does the resize keep the existing values though? ... no its not ) anyway once this would be solved the implementation is necessarily nice :grimacing: but it is easily extendible (well of course the primal handling of that value has to be considered as well, proper restart + calling just once during the simulation in the first iteration  )

I am unsure if there are applications which need these containers to become huge... so maybe an additional thought on not making it unnecessarily slow should be spent.

There was the idea to put this whole logic in a separate class (as far as I remember and understood correctly). But this instance still works the initial hacked version and that is a good thing :) .... I might have made all possible mistakes along the way and permuted a few &* in the process :cry:

Reply:
@pcarruscag  Any  comments to the implementation would be much appreciated :) When it comes to commenting all the things and cleaning some things -> I'll do it at the end